### ISSUE TITLE
Preventing traffic drop during revisioned upgrade

### ISSUE THREAD

**Issue by chauhanshubham** [istio/istio#44047]
**Preventing traffic drop during revisioned upgrades**

I have a traditional revisioned setup with two revisioned control planes provisioned, with `stable` and `canary`.
I intend to upgrade my bookinfo application from revision `stable` to newer revision `canary`.

I have the following GW and VS in place (spec truncated for brevity)
```yaml
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  labels:
    istio.io/rev: stable
  name: bookinfo-gateway
  namespace: default
spec:
  selector:
    istio: ingressgateway
  servers:
  - hosts:
    - '*'
    port:
      name: http
      number: 80
      protocol: HTTP
```

```yaml
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  labels:
    istio.io/rev: stable
  name: bookinfo
  namespace: default
spec:
  gateways:
  - bookinfo-gateway
  hosts:
  - '*'
  http:
  - match:
    - uri:
        exact: /productpage
...
    route:
    - destination:
        host: productpage
        port:
          number: 9080
```

When setting up the `canary` revision via the following command
```
istioctl install  --set revision=canary
```

the gateway proxies are upgraded in-place automatically to the `canary` revision as expected, and the gateway proxy connects to `istiod-canary`.

Since the GW and VS are still pointing to the previous revision `stable`, `istiod-canary` cannot provide details about these configs to the gateway proxy. This leads to traffic drop
```bash
root@http-client:/# curl -s -I -HHost:httpbin.example.com "http://172.18.254.100:80/productpage"
HTTP/1.1 503 Service Unavailable
content-length: 145
content-type: text/plain
date: Tue, 21 Mar 2023 14:00:53 GMT
server: envoy
```

The only workaround would be to 
Duplicate all istio configs and point that to the intended upgraded revision `canary`, in this case the GW and VS.
This would mean double the number of istio configs.

**Describe alternatives you've considered**
As mentioned earlier the only workaround is to have a copy of the istio configurations for all revisions in the setup. This leads to duplicates and IMHO is not a great UX.

Could we consider watching for and filter out istio configs, that not just matches the istiod revision, but matches multiple revisions at once?
i.e. in the scenario above, the pilot crd-client watcher in `istiod-canary` is able to watch for and process istio configs with revision `stable` and `canary` both.
If `istiod-canary` propagates the `stable` information to `canary` proxies, the traffic would still work. Users can then update the existing GW and VS objects to point to revision `canary`. This should not lead to any traffic drop. The rollback should also be seamless.

**Affected product area (please put an X in all that apply)**

[ ] Ambient
[ ] Docs
[X] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[X] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**

**Comment by deva26** [comment]
This looks like a very reasonable ask and some of our key clients are blocked by this. @howardjohn, can we consider the suggestion proposed by @chauhanshubham? Or is there any standard recommendation or otherwise  from the community to handle this situation?

**Comment by howardjohn** [comment]
Just don't put `istio.io/rev` on the VS and Gateway config and it will apply to all revisions

**Comment by deva26** [comment]
> Just don't put `istio.io/rev` on the VS and Gateway config and it will apply to all revisions

Unfortunately, that solution does not work well for our clients. There clusters have several revisions of istiod with equal number of revision labels in the Istio configurations. Removing the istio.io/rev from the VS/Gateway configs would result in conflicts and would have to be limited to specific Istio revisions (as in the case of the upgrade of one revision to the other)

**Comment by chauhanshubham** [comment]
As @deva26  mentioned, this won't work for our clients. The environments have multiple istio revisions deployed and helps in revision based config isolation.
Removing `istio.io/rev` completely would lead to breaking this revision based isolation, by considering the configs as "global" configs, which they are not.

consider -
proxies in `rev1` and `rev1+` are isolated in terms of service discovery and config propagation, from `rev2` and `rev2+`
Where proxies are upgrading from `rev1` to `rev1+` and so on.

Note that this revision based service discovery isolation is also enforced by setting `outboundTrafficPolicy: REGISTRY_ONLY`

We were looking at an enhancement that looks something like this
https://github.com/istio/istio/compare/master...chauhanshubham:multi-revisions-istiod-watcher?expand=1

**Comment by chauhanshubham** [comment]
Right now, Pilot filters the CR objects by comparing `REVISION` PilotArg and the `istio.io/rev` label in CRs

The [PoC code](https://github.com/istio/istio/compare/master...chauhanshubham:multi-revisions-istiod-watcher?expand=1) shows an approach like this

![Untitled Diagram (36)](https://user-images.githubusercontent.com/44999138/230350262-dbe7a6fd-4694-4e11-a105-98163d66b3d8.jpg)


---

Alternatively this filtering can be done by a separate PilotArg - Label matching altogether as well 

![Untitled Diagram (37)](https://user-images.githubusercontent.com/44999138/230350649-e7a4caba-7273-4e3f-91b2-a29bda6b9f40.jpg)

**Comment by chauhanshubham** [comment]
Hi @howardjohn / @hzxuzhonghu could i please get your suggestions on this. 

IMO the second approach, that allows for multiple istio control planes to discover a subset of istio configs, which are not strictly bound to revisions, seems to be a plausible feature request?

**Comment by hzxuzhonghu** [comment]
>multiple istio revisions deployed

Why are multi revisions deployed, is it for multi tenancy?


>Alternatively this filtering can be done by a separate PilotArg - Label matching altogether as well

IMO this is OR semantic with revision matching. can you make use discovery selectors?

**Comment by hzxuzhonghu** [comment]
As you say discovery selector is working on namespace, while `istio.io/rev ` work on per CR. I think these two methods are enough for your case. Maybe labels selector on per CR can let you create CRs in one namespace, but it also requires explicitly label the CRs, which can be a breaking change. And making CRs in one namespace violates the isolation rule, i recommend isolate with namespace boundary.


>https://github.com/istio/istio/blob/master/pkg/kube/kubetypes/types.go#L54-L71 suggests that these are client side filters that means the memory consumption would be there with istiod. I am getting confused on why this would be the case. May you please help me understand this?

	LabelSelector and 	FieldSelector are both server side filtering.  ObjectFilter is clien-side filter

**Comment by kanurag94** [comment]
Thanks for your speedy reply

IIUC, if we use namespace boundary, we would have to create namespace separately for `config-stable` and `config-canary` duplicating the same configuration in both. Thus not a great UX.

I think something like [this proposed change](https://github.com/istio/istio/compare/master...chauhanshubham:multi-revisions-istiod-watcher?expand=1) shouldn't mean a breaking change. It would just allow multiple revisions/tags to be filtered out by istiod. Can a `OR` semantic work?

A CR which is not tied to a revision + no label selector => no filtering - watched by all
A CR which is tied to a revision + no label selector => watched by same revisioned istiod
A CR which is tied to a revision + a label selector => watched by same revisioned istiod + matching label selector

> And making CRs in one namespace violates the isolation rule, i recommend isolate with namespace boundary.

Currently all istiod versions are created in a single namespace `istio-system`. The configurations which are common to all the workloads do usually reside in `istio-system` namespace only. Thus along with namespace boundary, having ability to mark CR to be used by which istiod revisions can help.
Basically user becomes free on how to filter which objects to watch. This could be useful, let's say even in a namespace, when someone just wishes to mark a stale CR or during testing.
